### PR TITLE
8305955: Remove Visual C++ specific workaround in javac

### PR DESCRIPTION
**Pull request by TheShermanTanker**
Visual C++ no longer requires the use of the i64 literal syntax and instead recommends the use of LL instead, so we should remove this workaround in the JNIWriter (this also helps when users write Windows JNI code meant to be compiled with alternate compilers other than Visual C++)

---------
### Progress
- [ ] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [ ] Change requires CSR request [JDK-8308990](https://bugs.openjdk.org/browse/JDK-8308990) to be approved

### Issues
 * [JDK-8305955](https://bugs.openjdk.org/browse/JDK-8305955): Remove Visual C++ specific workaround in javac (**Bug** - P4)
 * [JDK-8308990](https://bugs.openjdk.org/browse/JDK-8308990): Remove Visual C++ specific workaround in javac (**CSR**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/13457/head:pull/13457` \
`$ git checkout pull/13457`

Update a local copy of the PR: \
`$ git checkout pull/13457` \
`$ git pull https://git.openjdk.org/jdk.git pull/13457/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 13457`

View PR using the GUI difftool: \
`$ git pr show -t 13457`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/13457.diff">https://git.openjdk.org/jdk/pull/13457.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/13457#issuecomment-1506822379)

### CODE REVIEW

**Comment by lahodaj** [comment]
I think I'd like to understand more about "Visual C++ no longer requires" - since when, how (un-)realistic is it that people would still be using a C++ compiler that would require the workaround, etc.

Because, as far as I know, the `JNIWriter` is used by users to generate the headers, and those users may be using various compilers, so I think we should verify we won't break the users.

**Comment by TheShermanTanker** [comment]
While concrete documentation guarantees a minimum version of at least Visual C++ 2015 which supports the LL suffix, searches from earlier suggest that it supported this literal specifier as early as versions 2008 - 2010, with one source dating back all the way to Visual C++ 2005! Currently though, users will always have to use MSVC with the current JNIWriter when on Windows, since it is the only supported compiler that supports i64 as a literal suffix

https://learn.microsoft.com/en-us/cpp/cpp/numeric-boolean-and-pointer-literals-cpp?view=msvc-140#integer-literals
https://learn.microsoft.com/en-us/cpp/c-language/c-integer-constants?view=msvc-140

https://cboard.cprogramming.com/c-programming/98707-formatting-16-bytes-hex-2.html
https://stackoverflow.com/questions/14588997/llu-bad-suffix-on-number
https://stackoverflow.com/questions/2835746/how-do-you-specify-a-64-bit-unsigned-int-const-0x8000000000000000-in-vs2008

**Comment by TheShermanTanker** [comment]
Should there be a CSR for this change?

**Comment by TheShermanTanker** [comment]
As for confirmation that this works on, at the very minimum, Visual C++ 2015: https://godbolt.org/z/nqb7qcE4Y

**Comment by TheShermanTanker** [comment]
Bumping @lahodaj?

**Comment by lahodaj** [comment]
Not sure if I am expert enough on this (just trying to find out if there's a potential we would break builds of users), but might be good to take this through the CSR, just to be sure.

**Comment by TheShermanTanker** [comment]
I've created the CSR in full, not sure what to do from here on though

**Comment by TheShermanTanker** [comment]
Going to page @vicente-romero-oracle for help

**Comment by djelinski** [comment]
FWIW, here's the official documentation for VS2008 stating that LL is a supported suffix:
https://learn.microsoft.com/en-us/previous-versions/visualstudio/visual-studio-2008/00a1awxf(v=vs.90)
I couldn't find a similar page for VS2005.

**Comment by TheShermanTanker** [comment]
Ah, that's good to know, now we'll just have to wait for the CSR review

**Comment by TheShermanTanker** [comment]
Bumping

**Comment by jddarcy** [comment]
> Ah, that's good to know, now we'll just have to wait for the CSR review

Note that the CSR for this issue is in "Draft" state, which means it is not on the list of CSRs being considered for review. See https://wiki.openjdk.org/display/csr/Main for details.

To request a review, move the CSR to either Proposed or Finalized; HTH.

**Comment by TheShermanTanker** [comment]
Oops, thanks for the heads up

**Comment by TheShermanTanker** [comment]
CSR moved to "Proposed"

**Comment by TheShermanTanker** [comment]
Release Note created as https://bugs.openjdk.org/browse/JDK-8309551



> @TheShermanTanker only [Reviewers](https://openjdk.org/bylaws#reviewer) can determine that a CSR is not needed.

...

**Comment by TheShermanTanker** [comment]
Anyone? I also need help in removing the CSR unfortunately

**Comment by TheShermanTanker** [comment]
Bumping...

**Comment by TheShermanTanker** [comment]
Bumping again...

**Comment by TheShermanTanker** [comment]
Bumping...

**Comment by TheShermanTanker** [comment]
@jonathan-gibbons ?

**Comment by TheShermanTanker** [comment]
@jddarcy @lahodaj Anyone..?

**Comment by TheShermanTanker** [comment]
...

**Comment by jddarcy** [comment]
Catching up on reviews...

For the behavioral change being proposed, I think a CSR is deserved (as well as a release note).

I concur with @jonathan-gibbons 's suggestion of providing some kind of configuration knob to get the hold behavior in case is matters in an unanticipated situation.

**Comment by TheShermanTanker** [comment]
Alright, I've reopened the CSR now. There is already a Release Note:


> Release Note created as https://bugs.openjdk.org/browse/JDK-8309551
> 
> > @TheShermanTanker only [Reviewers](https://openjdk.org/bylaws#reviewer) can determine that a CSR is not needed.
> 
> ...

File: src/jdk.compiler/share/classes/com/sun/tools/javac/jvm/JNIWriter.java [hunk 1]
```diff
@@ -271,8 +271,7 @@ static void printStaticDefines(PrintWriter out, Pair<ClassSymbol, VarSymbol> p)
                 valueStr = value.toString() + "L";
                 break;
             case LONG:
-                // Visual C++ supports the i64 suffix, not LL.
-                valueStr = value.toString() + ((isWindows) ? "i64" : "LL");
+                valueStr = value.toString() + "LL";
```
Comment (jonathan-gibbons) [inline src/jdk.compiler/share/classes/com/sun/tools/javac/jvm/JNIWriter.java:274]: `JNIWriter` has access to the `Context` and thus to options, including "backdoor" options.

I suggest adding a definition of a member variable, something like the following, modeled on `checkAll`
```
private boolean visualCPlusPlus;

visualCPlusPlus = options.isSet("java:visualCPlusPlus");
```

then change line 275 to:
```
valueStr = value.toString() + ((isWindows && visualCPlusPlus) ? "i64" : "LL");
```

Overall, the intent is that the do-nothing default is the proposed new behavior, but that the command-line option `-XDvisualCPlusPlus` can be specified to recover the current behavior.

**Comment by TheShermanTanker** [comment]
Don't do that Skara, I will return to this patch in due time

**Pull request closed without merging**